pkg/geo/geomfn/linestring.go

```diff
@@ -252,3 +253,98 @@ func removePoint(lineString *geom.LineString, index int) (*geom.LineString, erro
 
 	return lineString.SetCoords(coords)
 }
+
+// LineSubString returns a LineString being a substring by start and end
+func LineSubString(g geo.Geometry, start, end float64) (geo.Geometry, error) {
+
+	if start < 0 || start > 1 || end < 0 || end > 1 {
+		return g, pgerror.Newf(pgcode.InvalidParameterValue,
+			"start and and must be within 0 and 1")
+	}
+	if start > end {
+		return g, pgerror.Newf(pgcode.Internal,
+			"2nd arg must be smaller then 3rd arg")
+	}
+
+	if start == end {
+		return LineInterpolatePoints(g, start, false)
+	}
+
+	lineT, err := g.AsGeomT()
+	if err != nil {
+		return g, err
+	}
+	lineString, ok := lineT.(*geom.LineString)
+	if !ok {
+		return g, pgerror.Newf(pgcode.InvalidParameterValue,
+			"first parameter has to be of type LineString")
+	}
+
+	// flat line should be return point empty immediately,
+	if lineString.Length() == 0 {
+		return geo.MakeGeometryFromGeomT(geom.NewPointEmpty(geom.XY).SetSRID(lineString.SRID()))
+	}
+
+	var newFlatCoords []float64
+	startDistance, endDistance := start*lineString.Length(), end*lineString.Length()
+
+	for i := range lineString.Coords() {
+		currentDistance, err := geom.NewLineString(geom.XY).SetCoords(lineString.Coords()[0 : i+1])
```

Comment:
can you use `NewLineStringFlat(geom.XY, lineString.FlatCoords()[:lineString.Layout().Stride() * i])`? the coordinates may not be x and y.

can we also call this `currentLineString`, as it does not store a distance.

Reply:
for the `NewLineStringFlat(geom.XY, lineString.FlatCoords()[:lineString.Layout().Stride() * i])`.
I think we cant use it, I test it and all the test case fails.
I think the `linestring.Coords` has already implemented `stride`, and I did `[0 : i+1]` to make the `currentLineString` is having the same `coords` with the `lineString`. because if only use the first `coords` and current `coords` the distance will be different